### PR TITLE
GC: Handle delete manifests and row level delete files

### PR DESCRIPTION
**Pull request by ajantha-bhat**
- `ManifestFiles.readPaths` is meant only for data manifests. 
Need to use `ManifestFiles.open` (package private) which can handle both data and delete manifests.
- Add a nice testcase to gc delete files and delete manifests (this took most of the effort 😄)

Fixes https://github.com/projectnessie/nessie/issues/7477

### CODE REVIEW

File: gc/gc-iceberg/src/main/java/org/apache/iceberg/ManifestReaderUtil.java
Comment (ajantha-bhat) [reply gc/gc-iceberg/src/main/java/org/apache/iceberg/ManifestReaderUtil.java:29]: Hard to convince the community. I will give it a try. 

Since we don't want to wait till the new Iceberg release for this issue to be fixed, we can still go ahead with this fix.

Comment (dimas-b) [reply gc/gc-iceberg/src/main/java/org/apache/iceberg/ManifestReaderUtil.java:29]: `ManifestFiles.open()` is small and straight-forward. Why not inline it into Nessie code while we wait for the Iceberg PR?

Comment (ajantha-bhat) [reply gc/gc-iceberg/src/main/java/org/apache/iceberg/ManifestReaderUtil.java:29]: `Manifest.open()` is not really the problem. We can call its internal functions directly and no need of making it public.

`entry.file()` is the real problem

**Comment by ajantha-bhat** [comment]
New testcase passed but the existing `ITSparkIcebergNessieS3` > `roundTrips` has failed with below callstack.

```
org.apache.spark.SparkException at ITSparkIcebergNessieS3.java:253
    Caused by: java.nio.file.AccessDeniedException at ITSparkIcebergNessieS3.java:253
        Caused by: org.apache.hadoop.fs.s3a.auth.NoAuthWithAWSException at ITSparkIcebergNessieS3.java:253
            Caused by: com.amazonaws.SdkClientException at ITSparkIcebergNessieS3.java:253
```

Locally I reverted the spark version bump to 3.4. It can pass. 
Now I need to figure out why it is impacting Auth :D

File: gc/gc-iceberg/src/main/java/org/apache/iceberg/ManifestReaderUtil.java
Comment (marvin-roesch) [reply gc/gc-iceberg/src/main/java/org/apache/iceberg/ManifestReaderUtil.java:29]: I think a case could be made that `ManifestFiles.readPaths` should "just" work for delete files as well. I don't see a reason why it shouldn't (both data and delete manifests list paths, etc.). That would in some sense be a minimally invasive change on Iceberg's end and might be easier to convince the community on?

Comment (ajantha-bhat) [reply gc/gc-iceberg/src/main/java/org/apache/iceberg/ManifestReaderUtil.java:29]: After reviewing the callers of `ManifestFiles.readPaths`, it looks like Iceberg uses it for only expire snapshots with branches (ReachableFileCleanup). And it doesn't handle delete files :) 

So, we need to support `ManifestFiles.readPaths` properly and also handle delete files for that scenario. I will raise a separate PR at Iceberg side. But Iceberg release may take months. We can go ahead with current changes at Nessie side as a temporary solution.

Comment (ajantha-bhat) [reply gc/gc-iceberg/src/main/java/org/apache/iceberg/ManifestReaderUtil.java:29]: > We can probably call entry.file() via java reflection (since the method is public, even though the interface is not).
Not just `entry.file()` is package private. 
`ManifestReader#liveEntries()` which is used in `readPathsFromManifest` is also a package private method. The result of that method `ManifestEntry` is also a package private.  So, I cannot cast in our package after reflection. 

As mentioned above, I will work on long term solution at Iceberg repo. As a short term solution, we might have to go with iceberg package names in Nessie.

**Review by snazy — APPROVED** [review]

**Comment by ajantha-bhat** [comment]
@dimas-b: We are waiting for your approval on this.

File: gc/gc-iceberg/src/main/java/org/apache/iceberg/ManifestReaderUtil.java
Comment (dimas-b) [reply gc/gc-iceberg/src/main/java/org/apache/iceberg/ManifestReaderUtil.java:29]: If @snazy is ok with this, I'm ok :)

**Review by dimas-b — APPROVED** [review]

**Comment by ajantha-bhat** [comment]
Thanks for the approvals and the review. I don't have permission to merge. So, please merge it when you guys see this.

**Pull request merged**
